File: sdk-core/src/main/java/care/data4life/sdk/ApiService.java
/*
 * Copyright (c) 2020 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk;

import com.squareup.moshi.Moshi;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import care.data4life.auth.AuthorizationService;
import care.data4life.sdk.lang.D4LException;
import care.data4life.sdk.lang.D4LRuntimeException;
import care.data4life.sdk.network.Environment;
import care.data4life.sdk.network.IHCService;
import care.data4life.sdk.network.model.CommonKeyResponse;
import care.data4life.sdk.network.model.DocumentUploadResponse;
import care.data4life.sdk.network.model.EncryptedRecord;
import care.data4life.sdk.network.model.UserInfo;
import care.data4life.sdk.network.typeadapter.EncryptedKeyTypeAdapter;
import care.data4life.sdk.util.Base64;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import okhttp3.CertificatePinner;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;

// TODO Kotlin and internal
public final class ApiService {

    private static final String HEADER_ALIAS = "gc_alias";
    private static final String HEADER_ACCESS_TOKEN = "access_token";
    private static final String HEADER_AUTHORIZATION = "Authorization";
    private static final String HEADER_BASIC_AUTH = "basic_auth";
    private static final String HEADER_GC_SDK_VERSION = "GC-SDK-Version";
    private static final String HEADER_TOTAL_COUNT = "x-total-count";
    private static final String PARAM_FILE_NUMBER = "file_number";
    private static final String PARAM_TEK = "tek";
    private static final String FORMAT_BEARER_TOKEN = "Bearer %s";
    private static final String FORMAT_BASIC_AUTH = "Basic %s";
    private static final String FORMAT_ANDROID_CLIENT_NAME = "Android %s";
    private static final String MEDIA_TYPE_OCTET_STREAM = "application/octet-stream";
    private static final int HTTP_401_UNAUTHORIZED = 401;

    private IHCService service;


    private final String clientID;
    private final String clientSecret;
    private final AuthorizationService authService;
    private final Environment environment;
    private String platform;
    private final NetworkConnectivityService connectivityService;
    private OkHttpClient client;
    private final String clientName;
    private final boolean debug;
    private final String staticAccessToken;

    /**
     * Full constructor.
     *
     * If the staticToken parameter is set to null, the SDK will handle the full OAuth flow
     * and fetch  access and retrieval tokens, renewing the former as needed for later
     * requests.
     * If the a non-null staticToken is passed, the SDK will use this value as an access token.
     * In this case, it will not dynamical fetch or renew tokens.
     *
     * @param authService         AuthorizationService
     * @param environment         Deployment environment
     * @param clientID            Client ID
     * @param clientSecret        Client secret
     * @param platform            Usage platform (D4L, S4H)
     * @param connectivityService Connectivity service
     * @param clientName          Client name
     * @param staticAccessToken   Prefetched OAuth token - if not null, it will be used directly (no token renewal).
     * @param debug               Debug flag
     */
    ApiService(AuthorizationService authService,
               Environment environment,
               String clientID,
               String clientSecret,
               String platform,
               NetworkConnectivityService connectivityService,
               String clientName,
               byte[] staticAccessToken,
               boolean debug) {
        this.authService = authService;
        this.environment = environment;
        this.clientID = clientID;
        this.clientSecret = clientSecret;
        this.platform = platform;
        this.connectivityService = connectivityService;
        this.clientName = clientName;
        this.debug = debug;
        this.staticAccessToken = staticAccessToken == null ? null : new String(staticAccessToken);
        configureService();
    }

    /**
     * Convenience constructor for instances that handle the OAuth flow themselves.
     *
     * @param authService         AuthorizationService
     * @param environment         Deployment environment
     * @param clientID            Client ID
     * @param clientSecret        Client secret
     * @param platform            Usage platform (D4L, S4H)
     * @param connectivityService Connectivity service
     * @param clientName          Client name
     * @param debug               Debug flag
     */
    ApiService(AuthorizationService authService,
               Environment environment,
               String clientID,
               String clientSecret,
               String platform,
               NetworkConnectivityService connectivityService,
               String clientName,
               boolean debug) {
        this(authService,
                environment,
                clientID,
                clientSecret,
                platform,
                connectivityService,
                clientName,
                null,
                debug);
    }

    private void configureService() {
        CertificatePinner certificatePinner = new CertificatePinner.Builder()
                .add(extractHostname(environment.getApiBaseURL(platform)), environment.getCertificatePin(platform))
                .build();

        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(debug ? HttpLoggingInterceptor.Level.HEADERS : HttpLoggingInterceptor.Level.NONE);

        Interceptor versionInterceptor = chain -> {
            Request request = chain.request();
            request = request.newBuilder()
                    .addHeader(HEADER_GC_SDK_VERSION, String.format(FORMAT_ANDROID_CLIENT_NAME, clientName))
                    .build();
            return chain.proceed(request);
        };

        // Pick authentication interceptor based on whether a static access token is used or not
        Interceptor authorizationInterceptor = this.staticAccessToken != null ? this::staticTokenIntercept : this::intercept;

        Interceptor retryInterceptor = chain -> {
            Request request = chain.request();
            Response response = null;
            try {
                response = chain.proceed(request);
            } catch (SocketTimeoutException exception) {
                if (connectivityService.isConnected()) {
                    response = chain.proceed(request);
                }
            }
            return response;
        };

        client = new OkHttpClient.Builder()
                .certificatePinner(certificatePinner)
                .addInterceptor(versionInterceptor)
                .addInterceptor(authorizationInterceptor)
                .addInterceptor(loggingInterceptor)
                .addInterceptor(retryInterceptor)
                .connectTimeout(2, TimeUnit.MINUTES)
                .readTimeout(2, TimeUnit.MINUTES)
                .writeTimeout(2, TimeUnit.MINUTES)
                .callTimeout(2, TimeUnit.MINUTES)
                .build();

        createService(environment.getApiBaseURL(platform));
    }

    private String extractHostname(String apiBaseURL) {
        return apiBaseURL.replaceFirst("https://", "");
    }

    OkHttpClient getClient() {
        return client;
    }

    private IHCService createService(String baseUrl) {

        Moshi moshi = new Moshi.Builder()
                .add(new EncryptedKeyTypeAdapter())
                .build();

        return service = new Retrofit.Builder()
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .baseUrl(baseUrl)
                .client(client)
                .build()
                .create(IHCService.class);

    }

    public Single<CommonKeyResponse> fetchCommonKey(String alias, String userId, String commonKeyId) {
        return service.fetchCommonKey(alias, userId, commonKeyId);
    }

    Completable uploadTagEncryptionKey(String alias, String userId, String encryptedKey) {
        Map<String, String> params = new HashMap<>();
        params.put(PARAM_TEK, encryptedKey);
        return service.uploadTagEncryptionKey(alias, userId, params);
    }

    Single<EncryptedRecord> createRecord(String alias, String userId, EncryptedRecord encryptedRecord) {
        return service.createRecord(alias, userId, encryptedRecord);
    }

    Observable<List<EncryptedRecord>> fetchRecords(String alias,
                                                   String userId,
                                                   String startDate,
                                                   String endDate,
                                                   Integer pageSize,
                                                   Integer offset,
                                                   List<String> tags) {
        return service.searchRecords(alias, userId, startDate, endDate, pageSize, offset, tags);
    }

    Completable deleteRecord(String alias, String recordId, String userId) {
        return service.deleteRecord(alias, userId, recordId);
    }

    Single<EncryptedRecord> fetchRecord(String alias, String userId, String recordId) {
        return service.fetchRecord(alias, userId, recordId);
    }

    Single<EncryptedRecord> updateRecord(String alias,
                                         String userId,
                                         String recordId,
                                         EncryptedRecord encryptedRecord) {
        return service.updateRecord(alias, userId, recordId, encryptedRecord);
    }

    // TODO remove public
    public Single<String> uploadDocument(String alias,
                                  String userId,
                                  byte[] encryptedAttachment) {
        return service.uploadDocument(
                alias, userId,
                RequestBody.create(MediaType.parse(MEDIA_TYPE_OCTET_STREAM), encryptedAttachment)
        ).map(DocumentUploadResponse::getDocumentId);
    }

    // TODO remove public
    public Single<byte[]> downloadDocument(String alias,
                                    String userId,
                                    String documentId) {
        return service.downloadDocument(alias, userId, documentId)
                .map(ResponseBody::bytes);
    }

    // TODO remove public
    public Single<Boolean> deleteDocument(String alias,
                                   String userId,
                                   String documentId) {
        // network request doesn't has a response except the HTTP 204
        // on success the method will always return `true`
        return service.deleteDocument(alias, userId, documentId).map(it -> true);
    }

    Single<Integer> getCount(String alias, String userId, List<String> tags) {
        return service
                .getRecordsHeader(alias, userId, tags)
                .map(response -> Integer.parseInt(response.headers().get(HEADER_TOTAL_COUNT)));
    }

    public Single<UserInfo> fetchUserInfo(String alias) {
        return service
                .fetchUserInfo(alias)
                .subscribeOn(Schedulers.io());
    }

    /**
     * Carry out needed logout actions.
     * <p>
     * When using refresh token, this will revoke the OAuth access.
     * When using a static access token, nothing will be done.
     *
     * @param alias Alias
     * @return Completable
     */
    public Completable logout(String alias) {
        if (this.staticAccessToken != null) {
            throw new D4LRuntimeException("Cannot log out when using a static access token!");
        }
        return Single
                .fromCallable(() -> authService.getRefreshToken(alias))
                .flatMapCompletable(token -> service.logout(alias, token));
    }

    /**
     * Interceptor that attaches an authorization header to a request.
     * <p>
     * The authorization can be basic auth or OAuth. In the OAuth case, the
     * interceptor will try the request snd if it comes back with a status code
     * 401 (unauthorized), it will update the OAuth access token using the
     * refresh token.
     *
     * @param chain OkHttp interceptor chain
     * @return OkHttp response
     * @throws IOException
     */
    private Response intercept(Interceptor.Chain chain) throws IOException {
        Request request = chain.request();
        String alias = request.header(HEADER_ALIAS);
        String authHeader = request.headers().get(HEADER_AUTHORIZATION);
        if (authHeader != null && authHeader.equals(HEADER_BASIC_AUTH)) {
            String auth = Base64.INSTANCE.encodeToString(clientID + ":" + clientSecret);
            request = request.newBuilder()
                    .removeHeader(HEADER_AUTHORIZATION)
                    .addHeader(HEADER_AUTHORIZATION, String.format(FORMAT_BASIC_AUTH, auth)).build();
        } else if (authHeader != null && authHeader.equals(HEADER_ACCESS_TOKEN)) {
            String tokenKey;
            try {
                tokenKey = authService.getAccessToken(alias);
            } catch (D4LException e) {
                return chain.proceed(request);
            }
            request = request.newBuilder()
                    .removeHeader(HEADER_AUTHORIZATION)
                    .removeHeader(HEADER_ALIAS)
                    .addHeader(HEADER_AUTHORIZATION, String.format(FORMAT_BEARER_TOKEN, tokenKey)).build();
            Response response = chain.proceed(request);
            if (response.code() == HTTP_401_UNAUTHORIZED) {
                try {
                    tokenKey = authService.refreshAccessToken(alias);
                } catch (D4LException e) {
                    authService.clear();
                    return response;
                }
                request = request.newBuilder()
                        .removeHeader(HEADER_AUTHORIZATION)
                        .addHeader(HEADER_AUTHORIZATION, String.format(FORMAT_BEARER_TOKEN, tokenKey)).build();
                response.close();
                return chain.proceed(request);
            }
            return response;
        }
        return chain.proceed(request);
    }

    /**
     * Interceptor that attaches an OAuth access token to a request.
     * <p>
     * This interceptor is used for the case where the SDK client does not
     * handle the OAuth flow itself and merely gets an access token injected.
     * Accordingly this interceptor does not attempt to refresh the access token
     * if the request should fail.
     *
     * @param chain OkHttp interceptor chain
     * @return OkHttp response
     * @throws IOException
     */
    private Response staticTokenIntercept(Interceptor.Chain chain) throws IOException {
        Request request = chain.request();
        String alias = request.header(HEADER_ALIAS);
        String authHeader = request.headers().get(HEADER_AUTHORIZATION);
        request = request.newBuilder()
                .removeHeader(HEADER_AUTHORIZATION)
                .removeHeader(HEADER_ALIAS)
                .addHeader(HEADER_AUTHORIZATION, String.format(FORMAT_BEARER_TOKEN, this.staticAccessToken)).build();
        return chain.proceed(request);
    }
}

File: sdk-core/src/main/java/care/data4life/sdk/network/ApiService.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */
package care.data4life.sdk.network

import care.data4life.auth.AuthorizationContract
import care.data4life.sdk.NetworkConnectivityService
import care.data4life.sdk.lang.D4LException
import care.data4life.sdk.lang.D4LRuntimeException
import care.data4life.sdk.network.interceptors.VersionInterceptor
import care.data4life.sdk.network.model.CommonKeyResponse
import care.data4life.sdk.network.model.DocumentUploadResponse
import care.data4life.sdk.network.model.EncryptedRecord
import care.data4life.sdk.network.model.NetworkModelContract
import care.data4life.sdk.network.model.UserInfo
import care.data4life.sdk.network.model.VersionList
import care.data4life.sdk.network.typeadapter.EncryptedKeyTypeAdapter
import care.data4life.sdk.util.Base64.encodeToString
import com.squareup.moshi.Moshi
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import okhttp3.CertificatePinner
import okhttp3.HttpUrl
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import java.io.IOException
import java.net.SocketTimeoutException
import java.util.HashMap
import java.util.concurrent.TimeUnit

// TODO Kotlin and internal
class ApiService constructor(
    private val authService: AuthorizationContract.Service,
    private val environment: NetworkingContract.Environment,
    private val clientID: String,
    private val clientSecret: String,
    private val platform: String?,
    private val connectivityService: NetworkConnectivityService,
    private val clientName: NetworkingContract.Clients,
    private val clientVersion: String,
    staticAccessToken: ByteArray?,
    private val debug: Boolean
) : NetworkingContract.Service {
    private var service: IHCService? = null
    var client: OkHttpClient? = null
        private set
    private val staticAccessToken: String? = staticAccessToken?.let { String(it) }

    /**
     * Convenience constructor for instances that handle the OAuth flow themselves.
     *
     * @param authService         AuthorizationService
     * @param environment         Deployment environment
     * @param clientID            Client ID
     * @param clientSecret        Client secret
     * @param platform            Usage platform (D4L, S4H)
     * @param connectivityService Connectivity service
     * @param clientName          Client name
     * @param clientName          Client version
     * @param debug               Debug flag
     */
    constructor(
        authService: AuthorizationContract.Service,
        environment: NetworkingContract.Environment,
        clientID: String,
        clientSecret: String,
        platform: String?,
        connectivityService: NetworkConnectivityService,
        clientName: NetworkingContract.Clients,
        clientVersion: String,
        debug: Boolean
    ) : this(
        authService,
        environment,
        clientID,
        clientSecret,
        platform,
        connectivityService,
        clientName,
        clientVersion,
        null,
        debug
    )

    private fun configureService() {
        val certificatePinner: CertificatePinner = CertificatePinner.Builder()
            .add(
                extractHostname(environment.getApiBaseURL(platform!!)),
                environment.getCertificatePin(
                    platform
                )
            )
            .build()
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.setLevel(if (debug) HttpLoggingInterceptor.Level.HEADERS else HttpLoggingInterceptor.Level.NONE)

        // Pick authentication interceptor based on whether a static access token is used or not
        val authorizationInterceptor = if (staticAccessToken != null) ::staticTokenIntercept else ::intercept
        val retryInterceptor = { chain: Interceptor.Chain ->
            val request = chain.request()
            var response: Response? = null
            try {
                response = chain.proceed(request)
            } catch (exception: SocketTimeoutException) {
                if (connectivityService.isConnected) {
                    response = chain.proceed(request)
                }
            }
            response!!
        }
        client = OkHttpClient.Builder()
            .certificatePinner(certificatePinner)
            .addInterceptor(
                VersionInterceptor.getInstance(
                    Pair(clientName, clientVersion)
                )
            )
            .addInterceptor(authorizationInterceptor)
            .addInterceptor(loggingInterceptor)
            .addInterceptor(retryInterceptor)
            .connectTimeout(NetworkingContract.REQUEST_TIMEOUT, TimeUnit.MINUTES)
            .readTimeout(NetworkingContract.REQUEST_TIMEOUT, TimeUnit.MINUTES)
            .writeTimeout(NetworkingContract.REQUEST_TIMEOUT, TimeUnit.MINUTES)
            .callTimeout(NetworkingContract.REQUEST_TIMEOUT, TimeUnit.MINUTES)
            .build()
        createService(environment.getApiBaseURL(platform))
    }

    private fun extractHostname(apiBaseURL: String): String {
        return apiBaseURL.replaceFirst("https://".toRegex(), "")
    }

    private fun createService(baseUrl: String): IHCService {
        val moshi = Moshi.Builder()
            .add(EncryptedKeyTypeAdapter())
            .build()
        return Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(MoshiConverterFactory.create(moshi))
            .baseUrl(baseUrl)
            .client(client)
            .build()
            .create(IHCService::class.java).also { service = it }
    }

    override fun fetchCommonKey(
        alias: String,
        userId: String,
        commonKeyId: String
    ): Single<CommonKeyResponse> {
        return service!!.fetchCommonKey(alias, userId, commonKeyId)
    }

    // TODO: Is this method even in use?
    override fun uploadTagEncryptionKey(
        alias: String,
        userId: String,
        encryptedKey: String
    ): Completable {
        val params: MutableMap<String, String> = HashMap()
        params[NetworkingContract.PARAM_TEK] = encryptedKey
        return service!!.uploadTagEncryptionKey(alias, userId, params)
    }

    override fun createRecord(
        alias: String,
        userId: String,
        encryptedRecord: NetworkModelContract.EncryptedRecord
    ): Single<EncryptedRecord> {
        return service!!.createRecord(alias, userId, (encryptedRecord as EncryptedRecord))
    }

    override fun updateRecord(
        alias: String,
        userId: String,
        recordId: String,
        encryptedRecord: NetworkModelContract.EncryptedRecord
    ): Single<EncryptedRecord> {
        return service!!.updateRecord(alias, userId, recordId, (encryptedRecord as EncryptedRecord))
    }

    override fun fetchRecord(
        alias: String,
        userId: String,
        recordId: String
    ): Single<EncryptedRecord> {
        return service!!.fetchRecord(alias, userId, recordId)
    }

    override fun searchRecords(
        alias: String,
        userId: String,
        startDate: String?,
        endDate: String?,
        pageSize: Int,
        offset: Int,
        tags: NetworkingContract.SearchTags
    ): Observable<List<EncryptedRecord>> {
        return service!!.searchRecords(
            alias,
            userId,
            startDate,
            endDate,
            pageSize,
            offset,
            tags.tags
        )
    }

    override fun countRecord(
        alias: String,
        userId: String,
        tags: NetworkingContract.SearchTags
    ): Single<Int> {
        return service!!
            .getRecordsHeader(alias, userId, tags.tags)
            .map { response ->
                response.headers()[NetworkingContract.HEADER_TOTAL_COUNT]!!
                    .toInt()
            }
    }

    override fun deleteRecord(alias: String, userId: String, recordId: String): Completable {
        return service!!.deleteRecord(alias, userId, recordId)
    }

    override fun uploadDocument(
        alias: String,
        userId: String,
        encryptedAttachment: ByteArray
    ): Single<String> {
        return service!!.uploadDocument(
            alias, userId,
            RequestBody.create(
                NetworkingContract.MEDIA_TYPE_OCTET_STREAM.toMediaType(),
                encryptedAttachment
            )
        ).map(DocumentUploadResponse::documentId)
    }

    override fun downloadDocument(
        alias: String,
        userId: String,
        documentId: String
    ): Single<ByteArray> {
        return service!!.downloadDocument(alias, userId, documentId)
            .map { response -> response.bytes() }
    }

    override fun deleteDocument(
        alias: String,
        userId: String,
        documentId: String
    ): Single<Boolean> {
        // network request doesn't has a response except the HTTP 204
        // on success the method will always return `true`
        return service!!.deleteDocument(alias, userId, documentId).map { true }
    }

    override fun fetchUserInfo(alias: String): Single<UserInfo> {
        return service!!
            .fetchUserInfo(alias)
            .subscribeOn(Schedulers.io())
    }

    override fun fetchVersionInfo(): Single<VersionList> {
        return service!!
            .fetchVersionInfo()
            .subscribeOn(Schedulers.io())
    }

    /**
     * Carry out needed logout actions.
     *
     *
     * When using refresh token, this will revoke the OAuth access.
     * When using a static access token, nothing will be done.
     *
     * @param alias Alias
     * @return Completable
     */
    override fun logout(alias: String): Completable {
        if (staticAccessToken != null) {
            throw D4LRuntimeException("Cannot log out when using a static access token!")
        }
        return Single
            .fromCallable { authService.getRefreshToken(alias) }
            .flatMapCompletable { token: String? -> service!!.logout(alias, token!!) }
    }

    /**
     * Interceptor that attaches an authorization header to a request.
     *
     *
     * The authorization can be basic auth or OAuth. In the OAuth case, the
     * interceptor will try the request snd if it comes back with a status code
     * 401 (unauthorized), it will update the OAuth access token using the
     * refresh token.
     *
     * @param chain OkHttp interceptor chain
     * @return OkHttp response
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun intercept(chain: Interceptor.Chain): Response {
        var request: Request = chain.request()
        val alias = request.header(NetworkingContract.HEADER_ALIAS)
        val authHeader = request.headers[NetworkingContract.HEADER_AUTHORIZATION]
        if (authHeader != null && authHeader == NetworkingContract.HEADER_BASIC_AUTH) {
            val auth = encodeToString("$clientID:$clientSecret")
            request = request.newBuilder()
                .removeHeader(NetworkingContract.HEADER_AUTHORIZATION)
                .addHeader(
                    NetworkingContract.HEADER_AUTHORIZATION,
                    String.format(NetworkingContract.FORMAT_BASIC_AUTH, auth)
                ).build()
        } else if (authHeader != null && authHeader == NetworkingContract.HEADER_ACCESS_TOKEN) {
            var tokenKey: String
            tokenKey = try {
                authService.getAccessToken(alias!!)
            } catch (e: D4LException) {
                return chain.proceed(request)
            }
            request = request.newBuilder()
                .removeHeader(NetworkingContract.HEADER_AUTHORIZATION)
                .removeHeader(NetworkingContract.HEADER_ALIAS)
                .addHeader(
                    NetworkingContract.HEADER_AUTHORIZATION,
                    String.format(NetworkingContract.FORMAT_BEARER_TOKEN, tokenKey)
                ).build()
            val response: Response = chain.proceed(request)
            if (response.code == NetworkingContract.HTTP_401_UNAUTHORIZED) {
                tokenKey = try {
                    authService.refreshAccessToken(alias)
                } catch (e: D4LException) {
                    authService.clear()
                    return response
                }
                request = request.newBuilder()
                    .removeHeader(NetworkingContract.HEADER_AUTHORIZATION)
                    .addHeader(
                        NetworkingContract.HEADER_AUTHORIZATION,
                        String.format(NetworkingContract.FORMAT_BEARER_TOKEN, tokenKey)
                    ).build()
                response.close()
                return chain.proceed(request)
            }
            return response
        }
        return chain.proceed(request)
    }

    /**
     * Interceptor that attaches an OAuth access token to a request.
     *
     *
     * This interceptor is used for the case where the SDK client does not
     * handle the OAuth flow itself and merely gets an access token injected.
     * Accordingly this interceptor does not attempt to refresh the access token
     * if the request should fail.
     *
     * @param chain OkHttp interceptor chain
     * @return OkHttp response
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun staticTokenIntercept(chain: Interceptor.Chain): Response {
        var request: Request = chain.request()
        val alias = request.header(NetworkingContract.HEADER_ALIAS)
        val authHeader = request.headers[NetworkingContract.HEADER_AUTHORIZATION]
        request = request.newBuilder()
            .removeHeader(NetworkingContract.HEADER_AUTHORIZATION)
            .removeHeader(NetworkingContract.HEADER_ALIAS)
            .addHeader(
                NetworkingContract.HEADER_AUTHORIZATION,
                String.format(NetworkingContract.FORMAT_BEARER_TOKEN, staticAccessToken)
            ).build()
        return chain.proceed(request)
    }

    /**
     * Full constructor.
     *
     *
     * If the staticToken parameter is set to null, the SDK will handle the full OAuth flow
     * and fetch  access and retrieval tokens, renewing the former as needed for later
     * requests.
     * If the a non-null staticToken is passed, the SDK will use this value as an access token.
     * In this case, it will not dynamical fetch or renew tokens.
     *
     * @param authService         AuthorizationService
     * @param environment         Deployment environment
     * @param clientID            Client ID
     * @param clientSecret        Client secret
     * @param platform            Usage platform (D4L, S4H)
     * @param connectivityService Connectivity service
     * @param clientName          Client name
     * @param clientVersion         Client version
     * @param staticAccessToken   Prefetched OAuth token - if not null, it will be used directly (no token renewal).
     * @param debug               Debug flag
     */
    init {
        configureService()
    }

    // TODO: This is a test concern and will be removed soon
    fun resetService(client: OkHttpClient, baseUrl: HttpUrl) {
        this.client = client
        this.service = createService(baseUrl.toString())
    }
}
